nixos/modules/system/boot/loader/grub/grub.nix

```diff
@@ -699,7 +747,7 @@ in
 
       boot.loader.grub.extraPrepareConfig =
         concatStrings (mapAttrsToList (n: v: ''
-          ${pkgs.coreutils}/bin/cp -pf "${v}" "@bootPath@/${n}"
+          ${pkgs.coreutils}/bin/install -Dp "${v}" "${efi.efiSysMountPoint}/"${escapeShellArg n}
```

Comment:
@samueldr This reads like it will break `mirroredBoots`.

[`boot.loader.efi.efiSysMountPoint`](https://search.nixos.org/options?channel=23.11&show=boot.loader.efi.efiSysMountPoint&from=0&size=50&sort=relevance&type=packages&query=efiSysMountPoint) appears to be a variable that must not be read if `mirroredBoods` is used (because then, there may be multiple [`boot.loader.grub.mirroredBoots.*.efiSysMountPoint`](https://search.nixos.org/options?channel=23.11&show=boot.loader.efi.efiSysMountPoint&from=0&size=50&sort=relevance&type=packages&query=efiSysMountPoint)s that all need to be handled instead).

`efi.efiSysMountPoint` is apparently only read in 1 place: To initialise `boot.loader.grub.mirroredBoots.*.efiSysMountPoint` if `mirroredBoots` are not explicitly defined:

https://github.com/NixOS/nixpkgs/blob/dd37924974b9202f8226ed5d74a252a9785aedf8/nixos/modules/system/boot/loader/grub/grub.nix#L708-L709



But with this change, you're switching to read `efi.efiSysMountPoint` directly, unconditionally.

This seems to directly undo the intention by @Frostman, in the commit that last touched this line: https://github.com/NixOS/nixpkgs/commit/7144b9ac5417c6e74af41fd3c6cad8e20ffc369c

> Fix boot.loader.grub.extraFiles when used with mirroredBoots
>
> Substitute `@bootPath@` in `boot.loader.grub.extraPrepareConfig` script same way as it's done for `boot.loader.grub.extraEntries` option.

Was the `@bootPath@` usage actually wrong, or did you just care about the usage of `-D` and `escapeShellArg`?

If yes, `@bootPath@` should probably be added back.

Also, I think there should be a comment on `efi.efiSysMountPoint` to say that it should not be read, because code written against it will not work with `mirroredBoots`. Where "not work" means "break" if the value is != any of the `mirroredBoots` one, and otherwise mean "silently act on only one of the ESPs, thus removing the benefits of `mirroredBoots` (system may not come back if one of the mirrored devices breaks)".